pallets/deeper-node/src/lib.rs

```diff
@@ -77,79 +95,427 @@ decl_module! {
         #[weight = 10_000]
         pub fn unregister_device(origin) -> DispatchResult {
             let sender = ensure_signed(origin)?;
-            ensure!(<DeviceInfo<T>>::contains_key(sender.clone()), "device not registered!");
-            let node = <DeviceInfo<T>>::get(sender.clone());
-            let country = node.country;
-            let _ = Self::try_remove_server(&sender, country);
+            ensure!(<DeviceInfo<T>>::contains_key(&sender), "device not registered!");
+            let _ = Self::try_remove_server(&sender);
             <DeviceInfo<T>>::remove(sender.clone());
             T::Currency::unreserve(&sender,BalanceOf::<T>::from(MIN_LOCK_AMT));
             Self::deposit_event(RawEvent::UnregisterNode(sender));
             Ok(())
         }
 
         #[weight = 10_000]
-        pub fn register_server(origin, country: u16) -> DispatchResult {
+        pub fn register_server(origin, duration: T::BlockNumber) -> DispatchResult {
             let sender = ensure_signed(origin)?;
-            ensure!(country < 1024, "invalid country code");
-            ensure!(<DeviceInfo<T>>::contains_key(sender.clone()),
-                "sender device needs register first");
-
-            let _ = Self::try_add_server(&sender, country);
-
+            ensure!(<DeviceInfo<T>>::contains_key(&sender),
+                    "sender device needs register first");
+            let _ = Self::try_remove_server(&sender);
+            let _ = Self::try_add_server(&sender, duration);
             Ok(())
         }
 
         #[weight = 10_000]
-        pub fn unregister_server(origin, country: u16) -> DispatchResult {
+        pub fn unregister_server(origin) -> DispatchResult {
             let sender = ensure_signed(origin)?;
-            ensure!(country < 1024, "invalid country code");
-            ensure!(<DeviceInfo<T>>::contains_key(sender.clone()),
-                "sender device needs register first");
-            let _ = Self::try_remove_server(&sender, country);
+            ensure!(<DeviceInfo<T>>::contains_key(&sender),
+                    "sender device needs register first");
+            let _ = Self::try_remove_server(&sender);
             Ok(())
         }
     }
 }
 
 impl<T: Trait> Module<T> {
-    // try to remove an account from a country's server list if exists
-    fn try_remove_server(sender: &T::AccountId, country: u16) -> DispatchResult {
-        let mut server_list = <ServersByCountry<T>>::get(country);
+    // try to remove an account from country and region server lists if exists
+    fn try_remove_server(sender: &T::AccountId) -> DispatchResult {
+        let mut node = <DeviceInfo<T>>::get(&sender);
+        ensure!(RegionMap::contains_key(&node.country), "invalid country code");
+        let first_region = RegionMap::get(&node.country);
```

Comment:
if the map is initialized correctly, we don't have to check first and second region